src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -480,6 +480,89 @@ public String getCookie() {
             return getDisplayName();
         }
 
+        static String findThreadName(FlowNode flowNode){
+            ThreadNameAction tna = flowNode.getPersistentAction(ThreadNameAction.class);
+
+            if (tna != null) {
+                return tna.getThreadName();
+            }
+            return null;
+        }
+
+        static String findStageName(FlowNode flowNode){
+            LabelAction la = flowNode.getPersistentAction(LabelAction.class);
+
+            if (la != null) {
+                return la.getDisplayName();
+            }
+            return null;
+        }
+
+        /**
+         * Similar to {@link #getEnclosingLabel()}.
+         * However instead of returning the innermost label including labels inside node blocks this one
+         * concatenates all labels found outside the current (node) block
+         *
+         * As {@link FlowNode#getEnclosingBlocks()} will return the blocks sorted from inner to outer blocks
+         * this method will create a string like
+         * <code>#innerblock#outerblock for</code> for a script like
+         * <pre>
+         *     {@code
+         *     parallel(outerblock: {
+         *         stage('innerblock') {
+         *             node {
+         *                 // .. do something here
+         *             }
+         *         }
+         *     }
+         *     }
+         * </pre>
+         *
+         * */
+        private @CheckForNull String concatenateAllEnclosingLabels() {
+            if (!context.isReady()) {
+                return null;
+            }
+            FlowNode executorStepNode;
+            try (Timeout t = Timeout.limit(100, TimeUnit.MILLISECONDS)) {
+                executorStepNode = context.get(FlowNode.class);
+            } catch (Exception x) {
+                LOGGER.log(Level.FINE, null, x);
+                return null;
+            }
+
+            StringBuilder labelName = new StringBuilder();
+            for(FlowNode node: executorStepNode.getEnclosingBlocks()) {
+                String threadName = findThreadName(node);
+                String stageName = findStageName(node);
+                String currentLabelName = threadName != null ? threadName: stageName;
+                if (threadName != null && stageName != null) {
+                    LOGGER.log(FINE, "\"concatenateAllEnclosingLabels(): Got both thread: {0} and stage: {1}. Will take thread name ({0}).",
```

Comment:
IIRC `parallel` blocks also set `LabelAction`, so it would suffice to ignore `ThreadNameAction`.

Reply:
@jglick: That means we can rely on that? That'd simplify the code. 👍

Reply:
I can remember that for some reason I chose to use `getThreadName()` but I cannot recall. However this was in some other code where the exact name did matter. Think in this context it'd not matter at all.

Reply:
Done.